BUG: Fix code to create the PANDAS tiles dataset

fepegar opened the pull request:
Fixes #314 by setting `margin` to 0.
Fixes also #310, but I will close that once the dataset has been generated.
Related to https://github.com/Project-MONAI/MONAI/issues/4156.

The script to create the tiles dataset now uses more descriptive file names for the tiles, using `left`, `top`, `right` and `bottom` for the file names and the CSV columns, instead of `x` and `y`.

fepegar commented on hi-ml-histopathology/src/histopathology/datasets/panda_dataset.py:
```diff
@@ -115,9 +118,11 @@ def __call__(self, data: Dict) -> Dict:
-                               size=(scaled_bbox.w, scaled_bbox.h),
-                               level=self.level)
+        get_data_kwargs = dict(
+            location=(level0_bbox.y, level0_bbox.x),
```
Note the swapped x and y here!

vale-salvatelli replied:
Interesting we never realized we were swapping the coordinates - is it something we can test?

fepegar replied:
> Interesting we never realized we were swapping the coordinates - is it something we can test?

I guess we could write a test for this using a synthetic WSI. Can we write WSIs?

kenza-bouzid replied:
Yep, you can use [MockPandaSlidesGenerator](https://github.com/microsoft/hi-ml/blob/39911d217c919d8213ad36c9c776f69369d98509/hi-ml-histopathology/testhisto/testhisto/mocks/slides_generator.py#L24)

fepegar replied:
Thanks, @kenza-bouzid! I'll take a look.

fepegar replied:
I would prioritise merging this in case it fixes the issue reported by @harshita-s, then we can think of adding some tests if you like. Do you agree?

fepegar commented on hi-ml-histopathology/src/histopathology/preprocessing/create_panda_tiles_dataset.py:
```diff
@@ -60,109 +79,114 @@ def save_image(array_chw: np.ndarray, path: Path) -> PIL.Image:
+        TileKey.TILE_TOP.value: tile_box.y,
+        TileKey.TILE_RIGHT.value: tile_box.x + tile_box.w,
+        TileKey.TILE_BOTTOM.value: tile_box.y + tile_box.h,
+        'data_provider': slide_metadata['data_provider'],
+        'slide_isup_grade': slide_metadata['isup_grade'],
+        'slide_gleason_score': slide_metadata['gleason_score'],
```
Couldn't find Enums for these.

dccastro replied:
Correct; metadata fields are free-form and specific to each dataset. That said, they're listed in `PandaDataset.METADATA_COLUMNS`. Not sure it's worth using that here, though.

fepegar commented on hi-ml-histopathology/src/histopathology/preprocessing/create_panda_tiles_dataset.py:
```diff
@@ -217,14 +246,69 @@ def main(panda_dir: Union[str, Path], root_output_dir: Union[str, Path], level:
+        action="store_true",
+        default=False,
+    )
+    parser.add_argument(
+        "--filter-slide",  # process only this slide
+        type=str,
+        default="",
+    )
```
This was useful for debugging; I can remove it if necessary.

dccastro replied:
Should be fine to leave it in, IMO. I also had a similar param when developing this script.

dccastro commented on hi-ml-histopathology/src/histopathology/preprocessing/create_panda_tiles_dataset.py:
```diff
@@ -60,109 +79,114 @@ def save_image(array_chw: np.ndarray, path: Path) -> PIL.Image:
+                tile_boxes[i],
+                slide_dir,
+            )
+            tile_metadata['occupancy'] = occupancies[i]
```
[optional] Should these keys be from `TileKey`?

dccastro commented on hi-ml-histopathology/src/histopathology/preprocessing/create_panda_tiles_dataset.py:
```diff
@@ -60,109 +79,114 @@ def save_image(array_chw: np.ndarray, path: Path) -> PIL.Image:
+                slide_dir,
+            )
+            tile_metadata['occupancy'] = occupancies[i]
+            tile_metadata['image'] = os.path.join(slide_dir.name, tile_metadata['image'])
```
[optional] Just noticed this could be simplified to `slide_dir / tile_metadata['image']` (same below), if you felt like it.

fepegar replied:
Not really because `slide_dir` is an absolute path. But I've made some edits to avoid using `os`. Thanks!

dccastro replied:
Ah yes, this is to produce the relative paths that go in the CSV

dccastro commented on hi-ml-histopathology/src/histopathology/utils/naming.py:
```diff
@@ -32,6 +32,10 @@ class TileKey(str, Enum):
     SPLIT = 'split'
     TILE_X = 'tile_x'
     TILE_Y = 'tile_y'
+    TILE_LEFT = 'left'
```
Do you envision us keeping the `TILE_X/Y` keys in addition to these new ones? Or should we refactor them out of existence later on?

fepegar replied:
As @vale-salvatelli, I think we should deprecate them at some point.

dccastro commented on hi-ml-histopathology/src/histopathology/preprocessing/create_panda_tiles_dataset.py:
```diff
@@ -217,14 +246,69 @@ def main(panda_dir: Union[str, Path], root_output_dir: Union[str, Path], level:
-         overwrite=False)
+    parser = ArgumentParser()
+    parser.add_argument(
+        "--container-dir",
```
What do you mean by 'container' in this context? Why not pass the PANDA folder directly?

fepegar replied:
I meant "folder on which an Azure blob container is mounted". I can't remember why I switched to specifying the container, but I'll revert back.

dccastro commented on hi-ml-histopathology/src/histopathology/preprocessing/create_panda_tiles_dataset.py:
```diff
@@ -217,14 +246,69 @@ def main(panda_dir: Union[str, Path], root_output_dir: Union[str, Path], level:
+        default=1,
+    )
+    parser.add_argument(
+        "--size",
```
[optional] Maybe `--tile_size`, for added clarity and consistency?

dccastro commented on hi-ml-histopathology/src/histopathology/preprocessing/create_panda_tiles_dataset.py:
```diff
@@ -217,14 +246,69 @@ def main(panda_dir: Union[str, Path], root_output_dir: Union[str, Path], level:
+        default=64,
+    )
+    parser.add_argument(
+        "--occupancy",
```
[optional] How about `--occupancy-threshold` or `--min-occupancy`?

dccastro reviewed (approved):
Looks good to merge, just added some suggestions and clarification questions. Was also wondering if you considered generating a small metadata file in the root output folder with the main tiling parameters (level, tile size, margin, occupancy threshold, etc.)?

vale-salvatelli commented on hi-ml-histopathology/src/histopathology/preprocessing/create_panda_tiles_dataset.py:
```diff
@@ -60,109 +79,114 @@ def save_image(array_chw: np.ndarray, path: Path) -> PIL.Image:
+    tile_boxes = [Box(x, y, tile_size_scaled, tile_size_scaled) for x, y in abs_tile_locations]
 
-    return image_tiles, mask_tiles, abs_tile_locations, occupancies, n_discarded
+    return image_tiles, mask_tiles, tile_boxes, occupancies, num_discarded
```
shall we test this function? it contains a lot of logic around offset/scale

fepegar replied:
The idea of this PR was to write a quick fix for PANDAS. It turned out to be not that quick due to the bugs. A similar function is used in the more generic script `create_tiles_dataset`. Discussing with @dccastro, we thought we could revisit and merge these scripts if we ever needed to generate more tiles datasets, which might be unlikely as we're moving towards online cropping. Then, we would move this logic outside of the script and write tests to cover it. Does that make sense?

vale-salvatelli replied:
I would still be in favor of writing a test just for this - if something is still wrong in the logic there will be errors in the recreated tiles. I am OK to wait for tests that concerns future datasets

fepegar replied:
I've added a test.

vale-salvatelli commented on hi-ml-histopathology/src/histopathology/preprocessing/create_panda_tiles_dataset.py:
```diff
@@ -60,109 +79,114 @@ def save_image(array_chw: np.ndarray, path: Path) -> PIL.Image:
 def process_slide(sample: dict, level: int, margin: int, tile_size: int, occupancy_threshold: int,
-                  output_dir: Path, tile_progress: bool = False) -> None:
-    slide_id = sample['image_id']
+                  output_dir: Path, tile_progress: bool = False, filter_slide: str = '') -> None:
```
not sure I understand what filter_slide is used for? Should this be a list?

fepegar replied:
I used it for debugging. It's useful to generate a dataset of a single slide. I'll add a help message.

vale-salvatelli commented on hi-ml-histopathology/src/histopathology/preprocessing/create_panda_tiles_dataset.py:
```diff
@@ -60,109 +79,114 @@ def save_image(array_chw: np.ndarray, path: Path) -> PIL.Image:
-            warnings.warn(f"An error occurred while processing slide {slide_id}: {e}")
+            shutil.rmtree(slide_dir)
+            return
+        logging.info(f"Tiling slide {slide_id} ...")
```
The following block was in the try/except before. In the current version we assume the error can come only from loading the slide. If I remember correctly, we had errors also on saving. It might be safer to wrap also this block in a try/except.

fepegar replied:
I got no errors on saving. They might have come from the issue with `squeeze`. I have narrowed down the try/except blocks to identify why errors happen and properly decide what to do with them.

vale-salvatelli replied:
as long as you can create the full dataset without errors looks OK to me - separate try/except would have been OK to capture different errors

fepegar replied:
I've updated the code to make error catching more granular. There are 4 slides with an empty mask and around 100 slides without an associated mask. All those slides are currently being ignored. For these cases, we could resort to computing the mask ourselves, but that might be out of the scope of this PR.

vale-salvatelli commented on hi-ml-histopathology/src/histopathology/preprocessing/create_panda_tiles_dataset.py:
```diff
@@ -217,14 +246,69 @@ def main(panda_dir: Union[str, Path], root_output_dir: Union[str, Path], level:
+        "--container-dir",
+        type=str,
+        required=True,
+        help="Folder with the 'datasets' container",
```
I would provide an example

vale-salvatelli commented on hi-ml-histopathology/src/histopathology/preprocessing/loading.py:
```diff
@@ -99,12 +99,12 @@ def __call__(self, data: Dict) -> Dict:
 
         data[self.image_key], _ = self.reader.get_data(image_obj, location=origin, level=self.level,
-                                                       size=(scaled_bbox.w, scaled_bbox.h))
+                                                       size=(scaled_bbox.h, scaled_bbox.w))
```
Now I understand why the axis were swapped before - why the need to swap axis in this transform? 

I couldn't find a test for this transform, this could be something worth to add as the transform was used in more than one place (e.g.  for the creation of TCGA-PRAD https://github.com/microsoft/hi-ml/blob/main/hi-ml-histopathology/src/histopathology/preprocessing/create_tiles_dataset.py)

harshita-s replied:
Not sure if anything changes in https://github.com/microsoft/hi-ml/blob/main/hi-ml-histopathology/src/histopathology/preprocessing/create_tiles_dataset.py ? This was used for tiling other slide datasets

fepegar replied:
> Now I understand why the axis were swapped before - why the need to swap axis in this transform?

Because the reader expects (y, x) coordinates.

> I couldn't find a test for this transform, this could be something worth to add as the transform was used in more than one place (e.g. for the creation of TCGA-PRAD https://github.com/microsoft/hi-ml/blob/main/hi-ml-histopathology/src/histopathology/preprocessing/create_tiles_dataset.py)

As I said above, I would wait to see if we need to create more tiles datasets before we add tests. Maybe we can add an issue to the backlog.

fepegar replied:
@harshita-s I'm not sure what you mean; I haven't touched that file.

fepegar replied:
Some more context about the coordinate swapping:
- https://github.com/microsoft/hi-ml/issues/310

harshita-s replied:
> @harshita-s I'm not sure what you mean; I haven't touched that file.

Was just wondering if we need to change that file as well (Daniel and Valentina suggested that it should be changed but this will be addressed in a later PR)

fepegar replied:
Ok, thanks for clarifying!

dccastro replied:
@fepegar, I think @harshita-s is wondering whether your changes in this PR affect the other script (e.g. via `LoadROId`, as @vale-salvatelli mentioned). We may need to revisit `create_tiles_dataset.py` somewhat soon e.g. for TCGA-PRAD. I'd be in favour of merging this now and discussing separately how to harmonise the scripts going forward.

fepegar replied:
Sounds good!

vale-salvatelli replied:
I am OK with getting in delaying this test if @dccastro is going to refactor and look into TCGA-PRAD soon (before we forget about this change). Adding a warning or a TODO in `create_tiles_dataset` would not harm

vale-salvatelli commented on hi-ml-histopathology/src/histopathology/utils/naming.py:
```diff
@@ -32,6 +32,10 @@ class TileKey(str, Enum):
     SPLIT = 'split'
     TILE_X = 'tile_x'
     TILE_Y = 'tile_y'
```
Maybe add a TODO to remove TILE_X. TILE_Y once all the datasets have been recreated according to this new philosophy

fepegar commented on hi-ml-histopathology/src/histopathology/preprocessing/create_panda_tiles_dataset.py:
```diff
@@ -40,15 +56,18 @@ def select_tile(mask_tile: np.ndarray, occupancy_threshold: float) \
     occupancy = foreground_mask.mean(axis=(-2, -1))
-    return (occupancy > occupancy_threshold).squeeze(), occupancy.squeeze()
+    selected = occupancy > occupancy_threshold
+    return selected[:, 0], occupancy[:, 0]
```
Because squeezing removes all dimensions when only one tile is to be selected, which raises an error later on.

```python
In [1]: import numpy as np

In [2]: selected = np.zeros((128, 1))

In [3]: selected[:, 0].shape
Out[3]: (128,)

In [4]: selected.squeeze().shape
Out[4]: (128,)

In [5]: selected = np.zeros((1, 1))

In [6]: selected[:, 0].shape
Out[6]: (1,)

In [7]: selected.squeeze().shape
Out[7]: ()
```

vale-salvatelli replied:
fair enough

fepegar commented on hi-ml-histopathology/src/histopathology/preprocessing/create_panda_tiles_dataset.py:
```diff
@@ -40,15 +56,18 @@ def select_tile(mask_tile: np.ndarray, occupancy_threshold: float) \
+def get_tile_descriptor(tile_box: Box) -> str:
+    left, top = tile_box.x, tile_box.y
+    right, bottom = left + tile_box.w, top + tile_box.h
+    return f"left_{left:05d}_top_{top:05d}_right_{right:05d}_bottom_{bottom:05d}"
```
Yes.

fepegar commented:
> Looks good to merge, just added some suggestions and clarification questions. Was also wondering if you considered generating a small metadata file in the root output folder with the main tiling parameters (level, tile size, margin, occupancy threshold, etc.)?

Sorry, I had missed this comment. I think none of these are actually needed for our purposes. The main concern was the tile size, but that is now explicit thanks to the bounding boxes being included in the CSV.
But I've added a log file in which all this is recorded anyway. It starts by mentioning the command used to generate the dataset, so in theory that should be enough to recover this information, if ever needed.
What do you think?

```console
$ cat /tmp/PANDA_tiles/panda_tiles_level1_224/2022-04-25_09-43-49.log     
2022-04-25 09:43:49 fernando-comp-westeu root[33137] INFO Command: "/home/azureuser/git/RadiomicsNN/innereye-deeplearning/hi-ml/hi-ml-histopathology/src/histopathology/preprocessing/create_panda_tiles_dataset.py --panda-dir /home/azureuser/datasets/panda --output-dir /tmp/PANDA_tiles --overwrite --no-parallel --margin 0 --filter-slide 0005f7aaab2800f6170c399693a96917"
2022-04-25 09:43:49 fernando-comp-westeu root[33137] INFO Creating dataset of level-1 224x224 PANDA tiles at: /tmp/PANDA_tiles/panda_tiles_level1_224
2022-04-25 09:43:49 fernando-comp-westeu root[33137] INFO >>> Slide dir /tmp/PANDA_tiles/panda_tiles_level1_224/0005f7aaab2800f6170c399693a96917
2022-04-25 09:43:49 fernando-comp-westeu root[33137] INFO Loading slide 0005f7aaab2800f6170c399693a96917 ...
2022-04-25 09:43:50 fernando-comp-westeu root[33137] INFO Tiling slide 0005f7aaab2800f6170c399693a96917 ...
2022-04-25 09:43:50 fernando-comp-westeu root[33137] INFO Discarded 107/168 tiles (63.69 %)
[...]
```

dccastro commented:
> Sorry, I had missed this comment. I think none of these are actually needed for our purposes. The main concern was the tile size, but that is now explicit thanks to the bounding boxes being included in the CSV. But I've added a log file in which all this is recorded anyway. It starts by mentioning the command used to generate the dataset, so in theory that should be enough to recover this information, if ever needed. What do you think?
That's great; should be enough in case we need that info later.

fepegar commented:
Thank you all for reviewing!
